lib/core/src/Cardano/Wallet/DB/Store/Wallets/Store.hs

```diff
@@ -17,109 +19,155 @@ Implementation of a store for 'TxWalletsHistory'
+import qualified Data.Set as Set
 
--- | Store for a map of 'DeltaTxMetaHistory' of multiple different wallets.
-mkStoreWalletsMeta :: Store
+
+mkStoreWalletMetaWithSubmissions :: W.WalletId -> Store
         (SqlPersistT IO)
-        (DeltaMap W.WalletId DeltaTxMetaHistory)
-mkStoreWalletsMeta =
+        (DeltaWalletsMetaWithSubmissions)
+mkStoreWalletMetaWithSubmissions wid =
     Store
     { loadS = load
     , writeS = write
     , updateS = update
     }
   where
-    write reset = forM_ (Map.assocs reset) $ \(wid,metas)
-        -> writeS (mkStoreMetaTransactions wid) metas
-    update _ (Insert wid metas) =
+    write (metas,subs) = do
         writeS (mkStoreMetaTransactions wid) metas
+        writeS (mkStoreSubmissions wid) subs
+    update (metas, TxLocalSubmissionHistory subs) xda = do
+        localTxToBeRemoved <- case xda of
+          ChangeMeta da -> do
+            updateS (mkStoreMetaTransactions wid) undefined da
+            let TxMetaHistory metas' = apply da metas
+            pure $ Set.difference (Map.keysSet subs) (Map.keysSet metas')
+          ChangeSubmissions da ->
+            updateS (mkStoreSubmissions wid) undefined da $> mempty
+        -- the next operation is potentially already executed via cascade delete
+        -- but we enforce it to avoid that DB feature in the future
+        forM_ localTxToBeRemoved $ \txId ->
+                deleteWhere
+                    [ LocalTxSubmissionWalletId ==. wid
+                    , LocalTxSubmissionTxId ==. txId
+                    ]
+    load = runExceptT $ do
+        metas <- ExceptT $ loadS (mkStoreMetaTransactions wid)
+        subs <- ExceptT $ loadS (mkStoreSubmissions wid)
+        pure (metas, subs)
+
+-- | Store for 'WalletsMeta' of multiple different wallets.
+mkStoreWalletsMetaWithSubmissions :: Store
+        (SqlPersistT IO)
+        (DeltaMap W.WalletId DeltaWalletsMetaWithSubmissions)
+mkStoreWalletsMetaWithSubmissions =
+    Store
+    { loadS = load
+    , writeS = write
+    , updateS = update
+    }
+  where
+    write reset = forM_ (Map.assocs reset) $ \(wid, ms) ->
+        writeS (mkStoreWalletMetaWithSubmissions wid) ms
+    update :: Map W.WalletId (TxMetaHistory, TxLocalSubmissionHistory)
+        -> DeltaMap W.WalletId DeltaWalletsMetaWithSubmissions
+        -> SqlPersistT IO ()
+    update _ (Insert wid ms) = do
+        writeS (mkStoreWalletMetaWithSubmissions wid) ms
     update _ (Delete wid) = do
         deleteWhere [TxMetaWalletId ==. wid ]
-    update _ (Adjust wid da) =
-        updateS (mkStoreMetaTransactions wid) undefined da
-    load :: SqlPersistT
-            IO
-            (Either SomeException (Map.Map W.WalletId TxMetaHistory))
-    load = do
-        wids <- nub . fmap (view #txMetaWalletId . entityVal)
+        deleteWhere [LocalTxSubmissionWalletId ==. wid ]
+    update old (Adjust wid xda) =
+        updateS (mkStoreWalletMetaWithSubmissions wid)
+            (Map.findWithDefault mempty wid old) xda
```

Comment:
```suggestion
    update old (Adjust wid xda) =
        case Map.lookup wid old of
            Nothing -> pure ()
            Just old' -> 
                updateS (mkStoreWalletMetaWithSubmissions wid) old' xda
```

As much as I like the clever avoidance of a case expression, I think the semantics of `Adjust` — implemented here

https://github.com/input-output-hk/cardano-wallet/blob/54b3264cf5cff4b45cbbc25024d76c37cbb1ad65/lib/dbvar/src/Data/DeltaMap.hs#L38

— require us to do nothing if the key is not in the map. 🤔